pkg/openstack/loadbalancer_test.go

```diff
@@ -534,3 +536,173 @@ func Test_getListenerProtocol(t *testing.T) {
+		expectedLen          int
+		expectError          bool
+		checkNewMembersEmpty bool
+	}{
+		{
+			name:        "NodePort == 0",
+			nodes:       []*corev1.Node{node1, node2},
+			port:        corev1.ServicePort{NodePort: 0},
+			svcConf:     svcConf,
+			expectedLen: 0,
+			expectError: false,
+		},
+		{
+			name:        "ErrNoAddressFound happens for all nodes",
+			nodes:       []*corev1.Node{{}, {}},
+			port:        port,
+			svcConf:     svcConf,
+			expectedLen: 0,
+			expectError: false,
+		},
+		{
+			name:  "Valid nodes, canUseHTTPMonitor=false",
+			nodes: []*corev1.Node{node1, node2},
+			port:  corev1.ServicePort{NodePort: 8080},
+			svcConf: &serviceConfig{
+				preferredIPFamily:   corev1.IPv4Protocol,
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+				enableMonitor:       true,
+			},
+			expectedLen: 2,
+			expectError: false,
+		},
+		{
+			name:  "Valid nodes, canUseHTTPMonitor=true",
+			nodes: []*corev1.Node{node1, node2},
+			port:  corev1.ServicePort{NodePort: 8080},
+			svcConf: &serviceConfig{
+				preferredIPFamily:   corev1.IPv4Protocol,
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+				enableMonitor:       true,
+			},
+			expectedLen: 2,
+			expectError: false,
+		},
+		{
+			name:  "ErrNoAddressFound happens and a member is ignored",
+			nodes: []*corev1.Node{node1, node2},
+			port:  corev1.ServicePort{NodePort: 8080},
+			svcConf: &serviceConfig{
+				preferredIPFamily:   corev1.IPv4Protocol,
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+				enableMonitor:       false,
+			},
+			expectedLen: 1,
+			expectError: true, // Expect an error
+		},
+		{
+			name:        "Valid nodes, NodePort != 0, canUseHTTPMonitor=false",
+			nodes:       []*corev1.Node{node1, node2},
+			port:        port,
+			svcConf:     svcConf,
+			expectedLen: 2,
+			expectError: false,
+		},
+		{
+			name:  "Valid nodes, NodePort != 0, canUseHTTPMonitor=true",
+			nodes: []*corev1.Node{node1, node2},
+			port:  corev1.ServicePort{NodePort: 8080},
+			svcConf: &serviceConfig{
+				preferredIPFamily:   corev1.IPv4Protocol,
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+				enableMonitor:       true,
+			},
+			expectedLen: 2,
+			expectError: false, // No error expected
+		},
+		{
+			name:  "Invalid preferred IP family, fallback to default",
+			nodes: []*corev1.Node{node1, node2},
+			port:  port,
+			svcConf: &serviceConfig{
+				preferredIPFamily:   "invalid-family",
+				lbMemberSubnetID:    "subnet-12345-test",
+				healthCheckNodePort: 8081,
+			},
+			expectedLen: 2,
+			expectError: false,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			members, newMembers, err := lbaas.buildBatchUpdateMemberOpts(tc.port, tc.nodes, tc.svcConf)
+
+			if tc.expectError {
+				// Expect an error
+				if assert.EqualError(t, err, "Expected an error in test case: "+tc.name) {
```

Comment:
I understand that `if assert` is documented in the library, however I find it very confusing. What does it do exactly? I suppose that this code would only check that `members` and `newMembers` are nil IF the error matches with the expected error.

Provided that you want to check that `members` and `newMembers` are nil, why wouldn't you want check that in case the error is not the same as expected?

Honestly I would remove the `if` block and put the three assertions one after the other without conditionals. Unless I am missing something?